### PR TITLE
lesson 3

### PR DESCRIPTION
**Pull request by rwngwn**

### CODE REVIEW

File: lesson-3-sheduling-all-around/README.md [hunk 1]
```diff
@@ -0,0 +1,232 @@
+
+Scheduling
+==========
+
+default scheduler is desribed at: https://kubernetes.io/docs/concepts/scheduling-eviction/kube-scheduler/
+
+
+The scheduler is working in multiple phases
+1) finding a feasible se of nodes
```
Comment (stlaz) [inline lesson-3-sheduling-all-around/README.md]: s/se/set/

File: lesson-3-sheduling-all-around/README.md [hunk 2]
```diff
@@ -0,0 +1,232 @@
+
+Scheduling
+==========
+
+default scheduler is desribed at: https://kubernetes.io/docs/concepts/scheduling-eviction/kube-scheduler/
+
+
+The scheduler is working in multiple phases
+1) finding a feasible se of nodes 
+1) scrorig feasible nodes so we can decide where workload will be places
```
Comment (stlaz) [inline lesson-3-sheduling-all-around/README.md]: s/we/it/
s/places/placed/

File: lesson-3-sheduling-all-around/README.md [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+
+Why this is so important? Becuase it can make your application not behave correctly and lead to potential downtime. Imagine situations like:
+
+1) all application pods running on same node
```
Comment (stlaz) [inline lesson-3-sheduling-all-around/README.md]: **the** same node

File: lesson-3-sheduling-all-around/README.md [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+1) all application pods running on node on same physical host
```
Comment (stlaz) [inline lesson-3-sheduling-all-around/README.md]: on a node on *the* same

File: lesson-3-sheduling-all-around/README.md [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+1) pods which should be together are spread on multiple nodes
```
Comment (stlaz) [inline lesson-3-sheduling-all-around/README.md]: s/spread on/spread among/

File: lesson-3-sheduling-all-around/README.md [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+1) ...
+
+We can prevent this by concept likes affinities and anti-affinities. They are described in a docs at: https://docs.openshift.com/container-platform/4.10/nodes/scheduling/nodes-scheduler-node-affinity.html
```
Comment (stlaz) [inline lesson-3-sheduling-all-around/README.md:19]: Maybe drop another sentence explaining these briefly.

File: lesson-3-sheduling-all-around/README.md [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+
+
+As an example to practice this we will create a three pods:
```
Comment (stlaz) [inline lesson-3-sheduling-all-around/README.md:22]: "As a practice example" maybe :slightly_smiling_face:

File: lesson-3-sheduling-all-around/README.md [hunk 8]
```diff
@@ -0,0 +1,232 @@
...
+  name: p1
+  labels:
+    security: S1
+spec:
+  containers:
+  - name: security-s1
+    image: docker.io/ocpqe/hello-pod
+```
+
+- p2 - which has affinity to be run on same node as "p1" using the selector be scheduled on same node as the pods with label "security" set to "S1"
```
Comment (stlaz) [inline lesson-3-sheduling-all-around/README.md:38]: using a selector **to** be scheduled... on **the** same node

File: lesson-3-sheduling-all-around/README.md [hunk 9]
```diff
@@ -0,0 +1,232 @@
...
+      requiredDuringSchedulingIgnoredDuringExecution:
+      - labelSelector:
+          matchExpressions:
+          - key: security
+            operator: In
+            values:
+            - S1
+```
+
+- p2 - which has antiAffinity to NOT be run on same node as "p1" using the selector be scheduled on same node as the pods with label "security" set to "S1"
```
Comment (stlaz) [inline lesson-3-sheduling-all-around/README.md:62]: p3
on *the* same node :slightly_smiling_face:

File: lesson-3-sheduling-all-around/README.md [hunk 10]
```diff
@@ -0,0 +1,232 @@
...
+
+If you are interested more in jsonpath you can find the docs at https://kubernetes.io/docs/reference/kubectl/jsonpath/
+
+
+To cleanup the environment we will run
+``` bash
+oc delete pod p1 p2 p3 p4
+```
+
+You can use these selector for more stuff. If you propagate your server placements, availability zones or HW equipment to lables. But be aware of bad resource utilization which can come with a complex rules and can even end up with upgrade issues.
```
Comment (stlaz) [inline lesson-3-sheduling-all-around/README.md]: s/these selector
 - this selector
 - these selectors

Not sure which option you meant.

Comment (stlaz) [reply lesson-3-sheduling-all-around/README.md]: s/lables/labels/

File: lesson-3-sheduling-all-around/README.md [hunk 11]
```diff
@@ -0,0 +1,232 @@
...
+
+Also think about scalability. For example if you need two services to be running on the same node consider them being part of one pod - then they can scale easily together.
+
+Quality of Service
+==================
+
+One of the part of workload schediling is also working with Quality of Service to assure that we have enough resources
```
Comment (stlaz) [inline lesson-3-sheduling-all-around/README.md]: s/One of the part/One of the parts
s/schediling/scheduling

File: lesson-3-sheduling-all-around/README.md (hunk 10 above)
Comment (stlaz) [reply lesson-3-sheduling-all-around/README.md]: Maybe add an example to schedule pod to a specific, labeled node with a `nodeSelector`, or mention the `nodeSelector` pod field specifically.

File: lesson-3-sheduling-all-around/README.md [hunk 12]
```diff
@@ -0,0 +1,232 @@
...
+to fullfill the needs of our workload.
+
+TBD
+
+
+Custome Scheduler
```
Comment (stlaz) [inline lesson-3-sheduling-all-around/README.md]: s/custome/custom/

File: lesson-3-sheduling-all-around/README.md [hunk 13]
```diff
@@ -0,0 +1,232 @@
...
+=================
+
+Custom scheduler is a little bit overhead in a lot of cases. We are using it to prove that it is possible and to demonstrate how to access k8s api from pods. The complete source of the scheduler can be viewed in the `scheduler/app.py`. In this scenario we will:
```
Comment (stlaz) [inline lesson-3-sheduling-all-around/README.md:110]: s/overhead/overkill/ ?

File: lesson-3-sheduling-all-around/README.md [hunk 14]
```diff
@@ -0,0 +1,232 @@
...
+
+- setup OpenShift build for the scheduler app
+- build it
+- create service account and set th required permissions
```
Comment (stlaz) [inline lesson-3-sheduling-all-around/README.md:114]: s/th/the/

File: lesson-3-sheduling-all-around/README.md [hunk 15]
```diff
@@ -0,0 +1,232 @@
...
+- run scheduler in OpenShift pod
+- Schedule a new pod using this scheduler
```
Comment (stlaz) [inline lesson-3-sheduling-all-around/README.md:116]: all other items begin with a lowercase letter

File: lesson-3-sheduling-all-around/README.md [hunk 16]
```diff
@@ -0,0 +1,232 @@
...
+
+So let's start with creating a binary build. This will enable us to build container in OpenShift using files on local filesystem. So run:
+``` bash
+oc new-build --binary --name=sched
+```
+
+
+
+
+Then we ask OpenShift to build our container by uploading the contet of a "scheduler" directory. Before executing this please change "NAME_SPACE" text to the project/namespace you are running your pods in. And then execute:
```
Comment (stlaz) [inline lesson-3-sheduling-all-around/README.md]: Where's this NAME_SPEC text to be found?

File: lesson-3-sheduling-all-around/README.md [hunk 17]
```diff
@@ -0,0 +1,232 @@
...
+``` bash
+oc get builds -w
+```
+
+to see the log of a build you can run
+``` bash
+oc logs sched-1-build --follow
+```
+
+When the job in the state of "Complete" we can continue by creating the service account. IT doesn't have enough permissions - but we will fix this in next step by meking it cluster-admin.
```
Comment (stlaz) [inline lesson-3-sheduling-all-around/README.md]: s/IT/It

File: lesson-3-sheduling-all-around/README.md [hunk 18]
```diff
@@ -0,0 +1,232 @@
...
+
+``` bash
+oc create sa sched
+```
+
+Now please ask in google meet chat for adding permissions to your service account. If you are on your cluster you can do it by command:
+
+``` bash
+oc adm policy add-cluster-role-to-user cluster-admin -z sched
```
Comment (stlaz) [inline lesson-3-sheduling-all-around/README.md:152]: ohoho, we're going very bold in this demo

File: lesson-3-sheduling-all-around/README.md [hunk 19]
```diff
@@ -0,0 +1,232 @@
...
+```
+
+
+We are ready to deploy it into OpenShift cluster. So please apply this to manifest to K8s. Dont forget to change IMAGE_STREAM_URL placeholder to the url revelead by `oc get is` command.
```
Comment (stlaz) [inline lesson-3-sheduling-all-around/README.md]: s/apply this to/apply this

File: lesson-3-sheduling-all-around/README.md [hunk 20]
```diff
@@ -0,0 +1,232 @@
...
+  containers:
+  - name: sched
+    image: IMAGE_STREAM_URL
+EOF
+```
+Notice the serviceAccountName in specs - yes this is used to specify different service account for a pod.
+Please check that pod is running corectly by examining its status and log.
+
+
+Now we will try to schedule three pods. One will be scheduled by default scheduler, one by ours and one will end up with a pending state.
```
Comment (stlaz) [inline lesson-3-sheduling-all-around/README.md]: by **the** default scheduler

**Review by stlaz — COMMENTED** [review]
a bit of a proofreading, some minor questions

File: lesson-3-sheduling-all-around/README.md [hunk 21]
```diff
@@ -0,0 +1,266 @@
...
+    image: docker.io/ocpqe/hello-pod
+```
+ 
+Now you can use `oc` command to double check if everything is running correctly.
+
+``` bash
+oc get pods -o jsonpath='{range .items[*]}{.metadata.name}{"\t"}{.spec.nodeName}{"\n"}'
+```
+
+As a task for yourself please create pod "p4" which will be scheduled on same node as pod "p3".
```
Comment (TomasTomecek) [inline lesson-3-sheduling-all-around/README.md:91]: nice, that's a good task for them

I hope we'll have multiple nodes in our cluster :D

File: lesson-3-sheduling-all-around/README.md [hunk 22]
```diff
@@ -0,0 +1,266 @@
...
+- setup OpenShift build for the scheduler app
+- build it
+- create service account and set th required permissions
+- run scheduler in OpenShift pod
+- Schedule a new pod using this scheduler
+
+
+So let's start with creating a binary build. This will enable us to build container in OpenShift using files on local filesystem. So run:
+``` bash
+oc new-build --binary --name=sched
```
Comment (TomasTomecek) [inline lesson-3-sheduling-all-around/README.md:121]: what a nonchalant way of introducing the attendees to a new openshift object :D

I can imagine there will be followup questions to builds

File: lesson-3-sheduling-all-around/README.md (hunk 18 above)
Comment (TomasTomecek) [reply lesson-3-sheduling-all-around/README.md:152]: exactly my thoughts! :O

File: lesson-3-sheduling-all-around/README.md [hunk 23]
```diff
@@ -0,0 +1,266 @@
...
+
+Second pod:
+``` yaml
+cat <<EOF | oc apply -f -
+apiVersion: v1
+kind: Pod
+metadata:
+  name: testsched2
+spec:
+  schedulerName: PrgContSched
```
Comment (TomasTomecek) [inline lesson-3-sheduling-all-around/README.md:199]: :)

File: lesson-3-sheduling-all-around/README.md (hunk 18 above)
Comment (stlaz) [reply lesson-3-sheduling-all-around/README.md:152]: Do we need cluster-admin, btw? It opens up doors to the nodes and possibly to the infrastructure running the cluster :eyes:

**Review by TomasTomecek — APPROVED** [review]

**Pull request merged**